compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2552,6 +2559,67 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             _ => {}
         }
     }
+
+    fn suggest_derive(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diagnostic,
+        trait_pred: ty::PolyTraitPredicate<'tcx>,
+    ) {
+        let Some(diagnostic_name) = self.tcx.get_diagnostic_name(trait_pred.def_id()) else {
+            return;
+        };
+        let (adt, substs) = match trait_pred.skip_binder().self_ty().kind() {
+            ty::Adt(adt, substs) if adt.did().is_local() => (adt, substs),
+            _ => return,
+        };
+        // Ensure all fields impl the trait.
+        if adt.all_fields().any(|field| {
```

Comment:
please move this below the can_derive check, since it is more expensive (checking traits on all types)

or possibly find out how to combine. but we should filter out the trait first.